library/x509_create.c

```diff
@@ -125,81 +136,352 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
+                {
+                    cur++;
+                    start++;
+                }
+
+                if( *cur == ',' )
+                {
+                    ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                    goto exit;
+                }
+                else if( *cur == '\"' )
+                {
+                    dn_cur_state = dn_in_quoted_value;
+                    start++;
+                }
+                else
+                {
+                    dn_cur_state = dn_in_unquoted_value;
+                    continue;
+                }
+            }
+            break;
 
-            // set tagType
-            cur->val.tag = attr_descr->default_tag;
-
-            while( c < end && *(c + 1) == ' ' )
-                c++;
-
-            s = c + 1;
-            in_tag = 1;
-        }
-
-        if( !in_tag && s != c + 1 )
-        {
-            *(d++) = *c;
+            /*
+             * Currently parsing a quoted value.
+             */
+            case dn_in_quoted_value:
+            {
+                if( *cur == '\"' && cur != end )
+                {
+
+                    if( value_len == 0 )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
+                        goto exit;
+                    }
+
+                    if( mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
+                                                       (unsigned char *) data,
+                                                       value_len ) == NULL )
+                    {
+                        ret = MBEDTLS_ERR_X509_ALLOC_FAILED;
+                        goto exit;
+                    }
+
+                    ( *head )->val.tag = attr_descr->default_tag;
+                    cur++;
+                    start = cur + 1;
+                    dn_cur_state = dn_after_quoted_value;
+                    continue;
+                }
+                else
+                {
+                    /*
+                     * Fill the current data pointer with the unescaped character.
+                     */
+                    *(d++) = *cur;
+                    value_len++;
+
+                    if( value_len == MBEDTLS_X509_MAX_DN_NAME_SIZE )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+                }
+            }
+            break;
+            /*
+             * We have finished parsing a quoted value.
+             * Skip all white spaces until we reach  the end or a comma.
+             * Any other character is an error.
+             */
+            case dn_after_quoted_value:
+            {
+                /*
+                 * Skip whitespaces.
+                 */
+                while( cur < end && *cur == ' ' )
+                {
+                    cur++;
+                }
+
+                if( *cur == ',' || cur == end )
```

Comment:
Same comment about the 1 byte overread. You might want to reverse the operands to || to fix this here.